### ISSUE TITLE
Negative loss encountered when using HingeEmbeddingLoss

### ISSUE THREAD

**Issue by stellaYi** [pytorch/pytorch#22965]
## 
May I know if the inputs to Hinge loss must be positive?

Encountered negative loss when I am training a  Multitask Deep Neural Network (MT-DNN) model done in pytorch.

If my understanding is correct, based on the formula here there should not be any negative loss? Since there is max{0,Δ−xn):
 
https://pytorch.org/docs/master/nn.html#torch.nn.HingeEmbeddingLoss

I get the MT-DNN model from:
https://github.com/namisan/mt-dnn

And i changed the loss function from F.cross_entropy to F.hinge_embedding_loss in mt_dnn/model.py.

I suspect the negative loss occurs only when y = 1, following is the observation:

extract from logs:
y: tensor([-1, -1], device='cuda:0')
logits: tensor([[0.9297, 1.1372],
        [1.1037, 0.9584]], device='cuda:0', grad_fn=<ThAddmmBackward>)
loss: tensor(0.0280, device='cuda:0', grad_fn=<DivBackward0>)
y: tensor([1, 1], device='cuda:0')
logits: tensor([[-40.4363, -37.7542],
        [-41.0224, -38.3677]], device='cuda:0', grad_fn=<ThAddmmBackward>)
loss: tensor(-39.3951, device='cuda:0', grad_fn=<DivBackward0>)

## To Reproduce

1. Change the loss function as mentioned above 
2. Run the finetune script in /scripts 
(note i am using my own finetune scripts, but mainly just path and dataset changes from the default one provided). Dataset is our own private dataset, not one of the GLUE dataset.  

Extracts of the log:
2019-07-17 05:19:25 [INFO] Total number of params: 109483778
2019-07-17 05:19:25 [WARNING] At epoch 0
2019-07-17 05:19:25 [INFO] Task [ 0] updates[     1] train loss[0.08631] remaining[0:26:52]
2019-07-17 05:20:13 [INFO] Task [ 0] updates[   500] train loss[0.42937] remaining[0:14:50]
2019-07-17 05:20:58 [INFO] Task [ 0] updates[  1000] train loss[0.28520] remaining[0:13:26]
2019-07-17 05:21:42 [INFO] Task [ 0] updates[  1500] train loss[0.13703] remaining[0:12:29]
2019-07-17 05:22:26 [INFO] Task [ 0] updates[  2000] train loss[0.00762] remaining[0:11:35]
2019-07-17 05:23:09 [INFO] Task [ 0] updates[  2500] train loss[-0.14033] remaining[0:10:45]
2019-07-17 05:23:52 [INFO] Task [ 0] updates[  3000] train loss[-0.31817] remaining[0:09:55]
2019-07-17 05:24:35 [INFO] Task [ 0] updates[  3500] train loss[-0.49544] remaining[0:09:08]
2019-07-17 05:25:19 [INFO] Task [ 0] updates[  4000] train loss[-0.69218] remaining[0:08:23]
2019-07-17 05:26:02 [INFO] Task [ 0] updates[  4500] train loss[-0.86834] remaining[0:07:38]
2019-07-17 05:26:46 [INFO] Task [ 0] updates[  5000] train loss[-1.05876] remaining[0:06:53]

## Expected behavior

loss reduces as training take place, but should not go below zero. 

## Environment

PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti
GPU 4: GeForce RTX 2080 Ti
GPU 5: GeForce RTX 2080 Ti
GPU 6: GeForce RTX 2080 Ti
GPU 7: GeForce RTX 2080 Ti

Nvidia driver version: 430.14
cuDNN version: /usr/local/cuda-9.0/lib64/libcudnn.so.7.6.0

Versions of relevant libraries:
[pip3] numpy==1.13.3
[conda] blas                      1.0                         mkl  
[conda] faiss-gpu                 1.5.3            py36h32c130d_0    pytorch
[conda] mkl                       2019.4                      243  
[conda] mkl_fft                   1.0.12           py36ha843d7b_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] pytorch-pretrained-bert   0.6.0                    pypi_0    pypi
[conda] torch                     0.4.1                    pypi_0    pypi

## Additional context

Did not encounter such issues when training with BERT model (using tensorflow) using the same dataset, I also modified it to hingeloss (using tensorflow library), loss is still positive.

**Comment by fmassa** [comment]
The implementation (and documentation) of `HingeEmbeddingLoss` does not enforce the loss to be positive if `target == 1`.

Note that `x` in `HingeEmbeddingLoss` is expected to be a measure of similarity. If your similarity function can be negative, then `HingeEmbeddingLoss` can also be negative.

Note that questions are better fitted to the discussion forums at https://discuss.pytorch.org , please continue your questions there (or in the `mt-dnn` repo) if you have further modeling questions.

**Comment by stellaYi** [comment]
Noted, thanks for the prompt response!
